### ISSUE TITLE
Selecting data for translation

### ISSUE THREAD

**Issue by OndrejVasicek** [yiisoft/yii2#11725]
### What steps will reproduce the problem?
* using the 'yii\i18n\DbMessageSource' for translations
* calling Yii:t() with sourceLanguage different  from language

### What is the expected result?
Some clean, performance-considering query. Something like this:

``` SQL
SELECT
	"t1"."message" AS "message",
	"t2"."translation" AS "translation"
FROM
	sp_source_message t1
JOIN sp_message t2 ON ("t1"."id" = "t2"."id")
WHERE
	"t1"."category" = 'app'
AND (
	"t2"."language" = 'cs-CZ'
	OR "t2"."language" = 'cs'
)
```

### What do you get instead?
``` SQL
(
	SELECT
		"t1"."message" AS "message",
		"t2"."translation" AS "translation"
	FROM
		"sp_source_message" "t1",
		"sp_message" "t2"
	WHERE
		("t1"."id" = "t2"."id")
	AND ("t1"."category" = 'app')
	AND ("t2"."language" = 'cs-CZ')
)
UNION ALL
	(
		SELECT
			"t1"."message" AS "message",
			"t2"."translation" AS "translation"
		FROM
			"sp_source_message" "t1",
			"sp_message" "t2"
		WHERE
			(
				("t1"."id" = "t2"."id")
				AND ("t1"."category" = 'app')
				AND ("t2"."language" = 'cs')
			)
		AND (
			"t2"."id" NOT IN (
				SELECT
					"id"
				FROM
					"sp_message"
				WHERE
					"language" = 'cs-CZ'
			)
		)
	)
```

### Additional info
Do we really need to seek for the ‘cs-CZ’ form of language? In the message table is only ‘cs’. Probably because my translation setting for console.
But even though – maybe I’m missing something, but the original query with subqueries seems to be unnecessary and highly ineffective. There is no need to explain what it does. It’s quite clear.

The other thing are indexes. There aren’t any. For small web, it’s not a problem, but I think this is a great example, where inserting and updating are so rare, that the ratio for read/insert(update) is extremely high.
So I wouldn’t hesitate to put indexes on language and category columns. I know how indexes works and I know that in many cases, they will be useless, because the number of same content in these columns will be same. But it’s still far better situation to have useless index on table, where reading is much higher, then situation we have large message table with loads of categories and no indexes. At least the category column is important. But again – from the principal perspective… :)
I know there is a caching option, but it just covers the problem.

| Q                | A
| ---------------- | ---
| Yii version      | 2.0.8
| PHP version      | 5.6.15
| Operating system | Win10

**Comment by OndrejVasicek** [comment]
Actually – there isn’t even an index for the foreign key between those two tables. I know mysql does it automatically and it’s needed (you can’t delete it), but Postgre doesn’t. Only on primary keys and unique constraints.

**Comment by samdark** [comment]
What is does is getting messages from `cs-CZ` and, if any messages is missing, getting it from `cs`. Your query produces different result since you're getting both if both exist.

**Comment by OndrejVasicek** [comment]
Ok, so by the complicated query you’re trying to find at least “cs” language, if there isn’t any “cs-CZ”, before you end up with the last option – showing the source language. Right?

My problem was, I didn’t wonder about option, where there are two versions of the same language. There is nothing like that in my Czech language, so I didn’t see the need.

Which leads me to some thoughts –I’ve read on some places, I should use the long version of my language, so instead of “cs” use “cs-CZ” in app setting. So I did. In the config for console message I use “cs”, but in the end, it doesn’t matter. The query is still complicated even if it’s sett to “cs-CZ”. But if I change the app language from “cs-CZ” to “cs”, the Yii is clever enough and look at the beautiful query now (even if I'm still wondering why Yii uses the old way of joining table and not the "join style"):

```sql
SELECT
	"t1"."message" AS "message",
	"t2"."translation" AS "translation"
FROM
	"sp_source_message" "t1",
	"sp_message" "t2"
WHERE
	("t1"."id" = "t2"."id")
AND ("t1"."category" = 'app')
AND ("t2"."language" = 'cs')
```

Well my question is – what’s the benefit of using “cs-CZ” if there isn’t any other version of “cs” language? Now it seems to be more reasonable to use only the “cs”.

But still – back to the overcomplicated query. I still thing it isn’t written good in performance wise, because the last subquery is executed for every row and it still selects the same. From my knowledge, subqueries are essential where you need to use some var from the query itself, like ID of the row or something. This isn’t the case. So in some terrible scenarios you can generate hundreds of not necessary subqueries.

I can see two solutions:

1. Do the first select first, store the ids into variable and then use it for excluding in the second part.
2. Use groupby and order like this 

```sql
… where "t2"."language" = 'cs-CZ' OR "t2"."language" = 'cs' group by (columns without t2.language) order by t2.language
```

Then due the group by you always get only one row per language, and the “cs” (shorter) will have bigger priority due the order statemen.

The question is. What’s more efficient. I believe the first one.

And the last thing – there is still open question about the indexes. What do you think about it?

**Comment by OndrejVasicek** [comment]
And there is still one another improvement, which crossed my mind. If I use more categories, sometimes more than one are used on the page. Is it even possible to first collect some info, which categories are gonna be used and then select them in one query?

**Comment by samdark** [comment]
> Ok, so by the complicated query you’re trying to find at least “cs” language, if there isn’t any “cs-CZ”, before you end up with the last option – showing the source language. Right?

Yes.

> even if I'm still wondering why Yii uses the old way of joining table and not the "join style"

That could be changed. Not critical though.

> Well my question is – what’s the benefit of using “cs-CZ” if there isn’t any other version of “cs” language? Now it seems to be more reasonable to use only the “cs”.

If you're translating only then none. CZ gives some hints to intl ICU about how to format currencies, dates etc. according to the country standards.

> I can see two solutions

If you can get EXPLAIN for both and probably measure performance difference for all three, that would be excellent.

> And the last thing – there is still open question about the indexes. What do you think about it?

If EXPLAIN will get better because of indexes, I'm open to adding these.

> And there is still one another improvement, which crossed my mind. If I use more categories, sometimes more than one are used on the page. Is it even possible to first collect some info, which categories are gonna be used and then select them in one query?

Yes, sounds like a good idea.

**Comment by OndrejVasicek** [comment]
Ok. I did my homework. Here are some specs:

| Q                | A
| ---------------- | ---
| DB engine | Postgre 9.4.5.
| Yii version | 2.0.8
| System | Win 10
| Hardware | Local PC with SSD disc; QuadCore 4,3Ghz; 16GB RAM

Test sourceMessage table has 19100 entries. 5 different categories – 10 000, 5 000, 3 000, 1 000, 100. Table messages has 5 languages, which is almost 100k entries in total. 

This of course isn’t real life example. But it can help see the results more clearly. On the other hand, the speed on busy or shared server without SSD could be also rapidly slower.
Every single query was executed right after PostgreSQL restart. So no internal cache was applied (tested). 5 runs for every query combination to see the range.

**Queries:**
* Query 1: the original one – see above
* Query 2: Splits the two selects. From first, takes the id, which uses in the second instead of the subquery – see above.
* Query2: Using the group by with orderby – see above.

Queries with EXPLAIN ANALYZE results: [yii2-messages-queries.zip](https://github.com/yiisoft/yii2/files/316427/yii2-messages-queries.zip)

Here are some results. I’ve compared all 3 queries on indexed and non-indexed tables. I’ve also done variations for the biggest category and smallest, because there are some big differences. Not only the amount of data varies, but even the internal optimizer of PostgreSQL changes the complexity of internal query as you can see in EXPLAIN. But it can vary a lot depending on db engine.

**Tables without Index:** 

* **Query 1**
  * Smallest category: 33 – 37 ms
  * Biggest category: 58 – 72 ms
* **Query 2**
  * Smallest category: 
    1. 7 - 11 ms
    2.	9 – 11 ms
  * Biggest category:
    1.	20 – 26 ms
    2.	40 – 44 ms
* **Query 3**
  *Smallest category: 9-21ms
  * Biggest category: 89 - 90 (ups :) )
  * Biggest category light (without GROUP BY and ORDER – we can easily manage it in PHP): 44 – 45 ms.

**Tables with index** (message.language, source_message.category):
* **Query 1**
  * Smallest category: 5 – 6 ms
  * Biggest category: 36 – 38 ms
* **Query 2**
  * Smallest category: 
    1.	5 - 10 ms
    2.	6 – 9 ms
  * Biggest category:
    1.	5 – 11 ms
    2.	37 – 42 ms
* **Query 3**
  * Smallest category: 5-7ms
  * Biggest category: 78 – 85 (ups again :) )
  * Biggest category light (without GROUP BY and ORDER – we can easily manage it in PHP): 35 – 41 ms.

**My conclusions?**

* The real life impact with any of these queries will be probably small. With cache turned on I’d easily turn my sight away.
* There really are some positive differences in my optimized queries, but also some drawbacks, depending on the size of category
* Buuuut everything fades out, when indexes come on stage.

**My suggestion?**

Let’s just apply the two indexes. It has only positives, no drawbacks.
What do you think about that?

**Comment by samdark** [comment]
I'm all for it. Do you want to make a pull request?

**Comment by OndrejVasicek** [comment]
Well I haven’t dig into migrations to much yet, but I guess add two indexes won’t be such a big deal. Ok then. I’ll take it in few days (won’t be here now).
What about the enhancement where Yii recognize what categories are gonna be used and only after that selects everything in one query. Do you think it’s worth it?

**Comment by samdark** [comment]
It complicates things a bit while not giving us much improvement (or am I wrong?) so I think while indexes are certainly an improvement, this change is questionable.

**Comment by OndrejVasicek** [comment]
Well there is my small try to do my first proper PR via this workflow - https://github.com/yiisoft/yii2/blob/master/docs/internals/git-workflow.md. Hope it helps a bit, because it’s not complete. There are some special SQL files for other SQL languages which I don’t know.

> It complicates things a bit while not giving us much improvement (or am I wrong?) so I think while indexes are certainly an improvement, this change is questionable.

I have no idea how complicated it could be, but if it’s more than “a bit”, you’re probably right - indexes and caching should be ok here. But in principle, in large projects with lots of categories, where this issue cumulates, it could help. So if there is some box for issue with low priority, you know, the one which nobody visits, we could put it there.

**Comment by samdark** [comment]
I think we'll finish indexes pull request for other DBs, merge it and close the issue till another request for performance improvement pops up.

**Comment by OndrejVasicek** [comment]
Ok. What about the other databases? I could check their docs and make something, but can’t truly test it. I know it’s just index but still.

**Comment by samdark** [comment]
That would be great.

**Comment by OndrejVasicek** [comment]
I used some already crated queries in Yii and some docs and added other SQL languages. Hope it's correct, but again - someone with the db installed should test it.

Is there a need to do another PR or commit after PR is legit approach?
